plutus-core/plutus-ir/test/GeneratorSpec.hs

```diff
@@ -0,0 +1,271 @@
+-- we actually get something out. As the generators are supposed to generate reasonable stuff this is
+-- a test of the compiler. I think we
+--  1. Want this somewhere
+--  2. Don't want it here
+--  Where do we want something like this?
+--  Also, the code below is a giant hack to "conenct to" the compiler at the "right" place (as judged
+--  by us when we were ripping the compiler apart to extract something that did something reasonable-ish)
+--
+-- TODO: we want this property somewhere!
+-- compile :: Term TyName Name DefaultUni DefaultFun ()
+--         -> Either (CompileError DefaultUni DefaultFun) (CompiledCode a)
+-- compile _tm = either Left Right $ runQuoteT $ do
+--   -- Make sure that names are unique (that's not guaranteed by QuickCheck)
+--   tm <- rename _tm
+--   plcTcConfig <- PLC.getDefTypeCheckConfig PIR.noProvenance
+--   let hints = UPLC.InlineHints $ \a _ -> case a of
+--                 PIR.DatatypeComponent PIR.Destructor _ -> True
+--                 _                                      -> False
+--       pirCtx = PIR.toDefaultCompilationCtx plcTcConfig
+--              & set (PIR.ccOpts . PIR.coOptimize) True
+--              & set (PIR.ccOpts . PIR.coPedantic) False
+--              & set (PIR.ccOpts . PIR.coVerbose) False
+--              & set (PIR.ccOpts . PIR.coDebug) False
+--              & set (PIR.ccOpts . PIR.coMaxSimplifierIterations)
+--                       (PIR.defaultCompilationOpts ^. PIR.coMaxSimplifierIterations)
+--              & set PIR.ccTypeCheckConfig Nothing
+--       uplcSimplOpts = UPLC.defaultSimplifyOpts
+--             & set UPLC.soMaxSimplifierIterations (PIR.defaultCompilationOpts ^. PIR.coMaxSimplifierIterations)
+--             & set UPLC.soInlineHints hints
+--
+--   plcT <- flip runReaderT pirCtx $ PIR.compileReadableToPlc $ fmap Original tm
+--   plcTcError <- runExceptT @(PLC.Error _ _ _)
+--              $ UPLC.deBruijnTerm =<< UPLC.simplifyTerm uplcSimplOpts (UPLC.erase plcT)
+--   case plcTcError of
+--     Left _   -> error "wrong"
+--     Right cc -> return $ DeserializedCode (UPLC.Program () (PLC.defaultVersion ()) $ void cc) Nothing mempty
+--
+-- prop_compile :: Property
+-- prop_compile =
+--   forAllDoc "_,tm" genTypeAndTermNoHelp_ (shrinkTypedTerm mempty mempty) $ \ (_, tm) ->
+--   isRight $ compile tm
+
+generators :: TestNested
+generators = return $ testGroup "generators"
+  [ testProperty "prop_genKindCorrect"  $ withMaxSuccess 1000 prop_genKindCorrect
+  , testProperty "prop_shrinkTypeSound" $ prop_shrinkTypeSound
+  , testProperty "prop_genTypeCorrect"  $ withMaxSuccess 1000 prop_genTypeCorrect
+  , testProperty "prop_shrinkTermSound" $ withMaxSuccess 20 prop_shrinkTermSound
+  ]
+
+-- * Core properties for PIR generators
+
+-- | Check that the types we generate are kind-correct
+prop_genKindCorrect :: Property
+prop_genKindCorrect =
+  forAllDoc "ctx" genCtx (const []) $ \ ctx ->
```

Comment:
No shrinking because QuickCheck shrinking causes dependent generators to be re-run - you don't keep the "inner" counterexamples around. This should probably be explained in a comment here.

Reply:
This is one important place where Hedgehog and QuickCheck differ if I understand correctly.

Reply:
Maybe you want to write a longer Note about what you're doing about shrinking and reference it.